File: integration-test/java/quarkus-app/src/main/java/com/qualityminds/lazyval/integration/domain/OrderDate.java
package com.qualityminds.lazyval.integration.domain;

import com.qualityminds.lazyval.LazyValue;

import java.time.LocalDate;

@LazyValue
public record OrderDate(LocalDate date) {
    public OrderDate {
        if(date.isAfter(LocalDate.now())){
            throw new IllegalArgumentException("Order date must not be in the future");
        }
    }

    public static OrderDate now(){
        return new OrderDate(LocalDate.now());
    }
}

File: integration-test/kotlin/jee/src/main/kotlin/com/qualityminds/lazyval/integration/domain/OrderDate.kt
package com.qualityminds.lazyval.integration.domain

import com.qualityminds.lazyval.LazyValue
import java.time.LocalDate

@LazyValue
data class OrderDate(val value: LocalDate) {
    init {
        require(!value.isAfter(LocalDate.now())) { "Order date must not be in the future" }
    }

    companion object {
        @JvmStatic
        fun now(): OrderDate = OrderDate(LocalDate.now())
    }
}
